src/contribute/testing.md

```diff
@@ -371,3 +371,135 @@ RunnableOnService tests are tests built to use the Beam TestPipeline class,
+    2. Addition of `PTransform`s  after the pipeline has already run.
+
+  Abandoned node detection is *automatically enabled* when a real pipeline 
+  runner (i.e. not a `CrashingRunner`) and/or a 
+  `@NeedsRunner` / `@RunnableOnService` annotation are detected. 
+  Consider the following test:
+  ```java
+  @Rule
+  public final transient TestPipeline pipeline = TestPipeline.create();
+  
+  @Test
+  @Category(NeedsRunner.class)
+  public void myPipelineTest() throws Exception {
+
+        final PCollection<String> pCollection = 
+          pipeline
+            .apply("Create", Create.of(WORDS).withCoder(StringUtf8Coder.of()))
+            .apply(
+                "Map1",
+                MapElements.via(
+                    new SimpleFunction<String, String>() {
+
+                      @Override
+                      public String apply(final String input) {
+                        return WHATEVER;
+                      }
+                    }));
+                    
+        PAssert.that(pCollection).containsInAnyOrder(WHATEVER);       
+
+		/* ERROR: pipeline.run() is missing, PAsserts are ineffective */
+      }
+  ```
+  
+  The `PAssert` at the end of this test method will not be executed, since 
+  `pipeline` is never run, making this test ineffective. If this test method 
+  is run using an actual pipeline runner, an exception will be thrown 
+  indicating that there was not `run()` invocation in the test.
+  
+  Exceptions that are thrown prior to executing a pipeline, will fail 
+  the test unless handled by an `ExpectedException` rule.
+  Consider the following test:
+  
+  ```java
+    @Rule
+    public final transient TestPipeline pipeline = TestPipeline.create();
+  
+    @Test
+    public void testReadingFailsTableDoesNotExist() throws Exception {
+      final String table = "TEST-TABLE";
+  
+      BigtableIO.Read read =
+          BigtableIO.read()
+              .withBigtableOptions(BIGTABLE_OPTIONS)
+              .withTableId(table)
+              .withBigtableService(service);
+  
+      // Exception will be thrown by read.validate() when read is applied.
+      thrown.expect(IllegalArgumentException.class);
+      thrown.expectMessage(String.format("Table %s does not exist", table));
+  
+      p.apply(read);
+   }
+  ```
+  The application of the `read` transform throws an exception, which is then 
+  handled by the `thrown` `ExpectedException` rule. 
+  In light of this exception, the fact this test has abandoned nodes 
+  (the `read` transform) does not play a role since the test fails before 
+  the pipeline would have been executed (had there been a `run()` statement). 
+   
+ * *Auto-add `run()`* (disabled by default)
+   A `TestPipeline` instance can be configured to auto-add a missing `run()` 
+   statement by setting `testPipeline.enableAutoRunIfMissing(true/false)`. 
+   If this feature is enabled, the snippet above will not throw an exception, 
```

Comment:
"snippet above" is a bit misleading since the next snippet up from this comment is about the ExpectedException stuff.